### ISSUE TITLE
[feature request] one_line and multi_lines cleaners.formatting

### ISSUE THREAD

**Issue by ghost** [federicocalendino/pysub-parser#6]
**Requested functions**
cleaners.formatting.one_line(subtitles):
"""Replaces all multi-line subtitles with a single-line sub."""

cleaners.formatting.multi_line(subtitles, max_len=40):
"""Refactors the line break of all subtitles, respecting the maximum character length per line defined in max_len"""

**Context:**
For automatic translation of subtitles, it is known that its application in subtitles with broken lines compromises the quality of the translation. In this way removing the line break, followed by the step of automatically translating the subtitle file, concluding with the application of the line break, increases the final quality of the auto-translated subtitle.

**Comment by fedecalendino** [comment]
Hi @apenasrr, would you be so kind to send me some examples for each one? 

Particularly, the second one. Im guessing that the line break should happen on spaces rather than in the middle of a word right?

**Comment by ghost** [comment]
yes. The break should happen on spaces.
Example:
"You ladies have a bet yourselves, but don't tell anyone else."

the 40º carac would be in: `do | n't`
but the break line should happen on: `but | don't`

Below is a rudimentary implementation of the function, which I prototyped in a project recently. It was a little test of concept. It needs a proper refactoring.
```
def sub_preprocess_breakline(path_file):

    subtitles = parser.parse(path_file)

    list_sub = []
    for sub in subtitles:
        list_lines = sub.lines
        sub.lines = format_multi_line(sub.lines, 40)
        list_sub.append(sub)
    return list_sub

def format_multi_line(list_lines, max_len=40):

    if len(list_lines) == 1:
        text_line = list_lines[0]
        new_list_line_1 = []
        new_list_line_2 = []
        first_line = True
        if len(text_line) > max_len:
            list_token = text_line.split(' ')
            for token in list_token:
                if len(' '.join(new_list_line_1)) + len(token) > max_len:
                    first_line = False
                if first_line:
                    new_list_line_1.append(token)
                else:
                    new_list_line_2.append(token)
        line_1 = ' '.join(new_list_line_1)
        line_2 = ' '.join(new_list_line_2)
        new_list_lines = [line_1, line_2]
        return new_list_lines
    else:
        return list_lines
```

I also advance a small snipet for the implementation of online formatting.
```
def format_one_line(subtitles):

    list_sub = []
    for sub in subtitles:
        # remove breakline
        sub.lines = [' '.join(sub.lines)]
        list_sub.append(sub)
    return list_sub
```
I recently did a job of auto-translating hundreds of subtitles and prototyping a framework to speed up the work and maximize the quality of this auto-translation.
I have other suggestions that can enrich the library (and that I have already implemented in my project), but I will leave them to propose in the near future.

Thanks for your project. Helped me a lot.

**Comment by fedecalendino** [comment]
Hi again @apenasrr,

for the first item, I think what you need is already there. By doing [`subtitle.text`](https://github.com/fedecalendino/pysub-parser/blob/main/pysubparser/classes/subtitle.py#L24) you'll get all the lines into one joined with a white space between them:

```python
for subtitle in subtitles:
    print(subtitle.text)

>>> Subtitle
>>> Multi line Subtitle


for subtitle in subtitles:
    print(subtitle.lines)

>>> ['Subtitle']
>>> ['Multi line', 'Subtitle']
```

As for the second part, I found a way to implement it using [`textwrap`](https://docs.python.org/3/library/textwrap.html). It's a quick test to see if it works, and so far seems all right. If it doesn't we can take a look at your snippet an see how can we refactor it!


```python
subtitle = Subtitle(
    index=0,
    lines=["You ladies have a bet yourselves, but don't tell anyone else."]
)

print(subtitle.lines)
>>> ["You ladies have a bet yourselves, but don't tell anyone else."]

clean_subtitle = list(
    wrap_text.clean([subtitle], width=40)
)[0]

print(clean_subtitle.lines)
>>> [
    'You ladies have a bet yourselves, but', 
    "don't tell anyone else.",
]
```


The new cleaner is included in the version **v1.7.0**. Let me know if this is what you were looking for! 

Cheers,
Fede.

**Comment by ghost** [comment]
"Let me know if this is what you were looking for!"
Yes! It solved it very clean. Thanks!

**Comment by ghost** [comment]
Hi! I believe there is still room for improvement. Just to register the idea, without abusing your good will, as I intend to implement it soon.

The new function applies a wraptext without paying attention to the aesthetics of the line break. There is room for improvement.

with max_len = 42

**example 1:**
"You ladies have a bet yourselves, but don't tell anyone else. Really, don't tell anyone."

- In actual version:
You ladies have a bet yourselves, but,
don't tell anyone else. Really, don't tell,
anyone.

- Possible nicer way to break line:
You ladies have a bet yourselves,
but don't tell anyone else.
Really, don't tell anyone.
 
**example 2:**
"word word word word word word word word word"
- In actual version:
word word word word word word word word
word

- Possible nicer way to break line:
word word word word word,
word word word word

Implementation:
```
def sub_wrap(sub_text, max_len):
    
    line_count = int(len_sub_text / max_len) + (len_sub_text % max_len > 0)
    mark_carac = int(len_sub_text/line_count)
    index_space = 0
    for index, carac in enumerate(sub_text[mark_carac:]):
        if carac == ' ':
            index_space = index
            break
    mark_break_line = mark_carac + index_space 
    return textwrap.wrap(sub_text, mark_break_line)
```

**Comment by ghost** [comment]
Using your lib, I was able to create a batch sub auto-translator. Thanks again!

**Batch Subs Translator**
https://gist.github.com/apenasrr/e2da717655ef0ac52d031fd8e410963e

**Comment by fedecalendino** [comment]
Amazing! Im so glad that you find this lib so useful 😁

**Comment by ghost** [comment]
@fedecalendino I did not want to open  new issue for that question so wanted to ask it here, does it delete karaokes too? while converting ass to srt.
since karaoke gives results like that 
```143
00:03:16,490 --> 00:03:17,890
me

144
00:03:17,050 --> 00:03:18,450
ga

145
00:03:17,310 --> 00:03:18,710
ri

146
00:03:17,660 --> 00:03:19,060
no

147
00:03:17,810 --> 00:03:19,210
so

148
00:03:18,360 --> 00:03:19,760
ra

149
00:03:18,530 --> 00:03:19,930
wo

150
00:03:20,690 --> 00:03:22,090
gu

151
00:03:21,050 --> 00:03:22,450
ta

152
00:03:21,270 --> 00:03:22,670
bi
```

**Comment by fedecalendino** [comment]
Hi @toprak can you send me some of the songs/ids that you were checking? 

I'll take a look.

**Comment by ghost** [comment]
its an anime subtitle so here, I want to delete karaokes too
https://we.tl/t-PVNfgD9gOe

**Comment by Ashmit12092000** [comment]
i'm facing an issue where it is parsing till second last line of srt file ,  this is my input srt 
1
00:00:03,400 --> 00:00:06,177
In this lesson, we're going to
be talking about finance. And

2
00:00:06,177 --> 00:00:10,009
one of the most important aspects
of finance is interest.

3
00:00:10,009 --> 00:00:13,655
When I go to a bank or some
other lending institution

4
00:00:13,655 --> 00:00:17,720
to borrow money, the bank is happy
to give me that money. But then I'm

5
00:00:17,900 --> 00:00:21,480
going to be paying the bank for the
privilege of using their money. And that

6
00:00:21,660 --> 00:00:26,440
amount of money that I pay the bank is
called interest. Likewise, if I put money

7
00:00:26,620 --> 00:00:31,220
in a savings account or I purchase a
certificate of deposit, the bank just

8
00:00:31,300 --> 00:00:35,800
doesn't put my money in a little box
and leave it there until later. They take

9
00:00:35,800 --> 00:00:40,822
my money and lend it to someone
else. So they are using my money.

10
00:00:40,822 --> 00:00:44,400
The bank has to pay me for the privilege
of using my money.

11
00:00:44,400 --> 00:00:48,700
Now what makes banks
profitable is the rate

12
00:00:48,700 --> 00:00:53,330
that they charge people to use the bank's
money is higher than the rate that they

13
00:00:53,510 --> 00:01:00,720
pay people like me to use my money. The
amount of interest that a person pays or

14
00:01:00,800 --> 00:01:06,640
earns is dependent on three things. It's
dependent on how much money is involved.

15
00:01:06,820 --> 00:01:11,300
It's dependent upon the rate of interest
being paid or the rate of interest being

16
00:01:11,480 --> 00:01:17,898
charged. And it's also dependent upon
how much time is involved. If I have

17
00:01:17,898 --> 00:01:22,730
a loan and I want to decrease the amount
of interest that I'm going to pay, then

18
00:01:22,800 --> 00:01:28,040
I'm either going to have to decrease how
much money I borrow, I'm going to have

19
00:01:28,220 --> 00:01:32,420
to borrow the money over a shorter period
of time, or I'm going to have to find a

20
00:01:32,600 --> 00:01:37,279
lending institution that charges a lower
interest rate. On the other hand, if I

21
00:01:37,279 --> 00:01:41,480
want to earn more interest on my
investment, I'm going to have to invest

22
00:01:41,480 --> 00:01:46,860
more money, leave the money in the
account for a longer period of time, or

23
00:01:46,860 --> 00:01:49,970
find an institution that will pay
me a higher interest rate.


But pysubparser subtitle output is :

0 > In this lesson, we're going to be talking about finance. And
1 > one of the most important aspects of finance is interest.
2 > When I go to a bank or some other lending institution
3 > to borrow money, the bank is happy to give me that money. But then I'm
4 > going to be paying the bank for the privilege of using their money. And that
5 > amount of money that I pay the bank is called interest. Likewise, if I put money
6 > in a savings account or I purchase a certificate of deposit, the bank just
7 > doesn't put my money in a little box and leave it there until later. They take
8 > my money and lend it to someone else. So they are using my money.
9 > The bank has to pay me for the privilege of using my money.
10 > Now what makes banks profitable is the rate
11 > that they charge people to use the bank's money is higher than the rate that they
12 > pay people like me to use my money. The amount of interest that a person pays or
13 > earns is dependent on three things. It's dependent on how much money is involved.
14 > It's dependent upon the rate of interest being paid or the rate of interest being
15 > charged. And it's also dependent upon how much time is involved. If I have
16 > a loan and I want to decrease the amount of interest that I'm going to pay, then
17 > I'm either going to have to decrease how much money I borrow, I'm going to have
18 > to borrow the money over a shorter period of time, or I'm going to have to find a
19 > lending institution that charges a lower interest rate. On the other hand, if I
20 > want to earn more interest on my investment, I'm going to have to invest
21 > more money, leave the money in the account for a longer period of time, or


And python code is:
`from pysubparser import parser

subtitle_file = "res.srt"
subtitles = parser.parse(subtitle_file,encoding="utf-8")

for subtitle in subtitles:
    print(subtitle)`

**Comment by FreeTymeKiyan** [comment]
facing the same issue as @Ashmit12092000

**Comment by FreeTymeKiyan** [comment]
The code in `srt.py` seems to require an additional blank line at the end of the srt file. Otherwise, the last line won't be yield to the result

https://github.com/fedecalendino/pysub-parser/blob/main/pysubparser/parsers/srt.py#L41C22-L41C22

**Comment by fedecalendino** [comment]
Hi @toprak @Ashmit12092000 and @FreeTymeKiyan !

Apologies for the late response, this issue definitively went under my radar. The bug was to a missing yield at the end of the parser.

I just published the version *v1.7.1* for this library with the fix. Here is a screenshot with @toprak's example and it looks to be good.

![image](https://github.com/fedecalendino/pysub-parser/assets/7383708/e4767583-a76f-43b6-a195-84672260b422)

Please let me know if this fixes your problem.

Cheers, Fede.
